### PR TITLE
schedule: record elapsed time of operator step

### PR DESCRIPTION
**Pull request by Connor1996**

### CODE REVIEW

File: server/schedule/operator.go [hunk 1]
```diff
@@ -181,8 +183,10 @@ func (o *Operator) Step(i int) OperatorStep {
 func (o *Operator) Check(region *core.RegionInfo) OperatorStep {
 	for step := atomic.LoadInt32(&o.currentStep); int(step) < len(o.steps); step++ {
 		if o.steps[int(step)].IsFinish(region) {
+			operatorStepDuration.WithLabelValues(reflect.TypeOf(o.steps[int(step)]).Name()).Observe(time.Since(o.stepTime).Seconds())
 			atomic.StoreInt32(&o.currentStep, step+1)
 		} else {
+			o.stepTime = time.Now()
```
Comment (nolouch) [inline server/schedule/operator.go]: if two times check when step not finished, it will update the start time.

Comment (disksing) [reply server/schedule/operator.go]: And the `stepTime` is not set up when creating the operator.

Comment (Connor1996) [reply server/schedule/operator.go]: @nolouch that is for purpose, we don't need care about the first check that step not finished finally. For the second check, it updates the start time, and then we can get the correct elapsed time of the second try.

File: server/schedule/operator.go [hunk 2]
```diff
@@ -181,7 +184,9 @@ func (o *Operator) Step(i int) OperatorStep {
 func (o *Operator) Check(region *core.RegionInfo) OperatorStep {
 	for step := atomic.LoadInt32(&o.currentStep); int(step) < len(o.steps); step++ {
 		if o.steps[int(step)].IsFinish(region) {
+			operatorStepDuration.WithLabelValues(reflect.TypeOf(o.steps[int(step)]).Name()).Observe(time.Since(o.stepTime).Seconds())
 			atomic.StoreInt32(&o.currentStep, step+1)
+			o.stepTime = time.Now()
```
Comment (disksing) [inline server/schedule/operator.go]: Any need to make it atomic?

**Comment by disksing** [comment]
LGTM. Don't forget to update grafana template.

File: server/schedule/operator.go [hunk 3]
```diff
@@ -181,7 +184,11 @@ func (o *Operator) Step(i int) OperatorStep {
 func (o *Operator) Check(region *core.RegionInfo) OperatorStep {
 	for step := atomic.LoadInt32(&o.currentStep); int(step) < len(o.steps); step++ {
 		if o.steps[int(step)].IsFinish(region) {
+			operatorStepDuration.WithLabelValues(reflect.TypeOf(o.steps[int(step)]).Name()).
+				Observe(time.Since(time.Unix(0, atomic.LoadInt64(&o.stepTime))).Seconds())
 			atomic.StoreInt32(&o.currentStep, step+1)
+			atomic.StoreInt64(&o.stepTime, time.Now().UnixNano())
+			time.Unix(0, 0)
```
Comment (nolouch) [inline server/schedule/operator.go]: need remove?

File: server/schedule/operator.go [hunk 4]
```diff
@@ -181,7 +184,11 @@ func (o *Operator) Step(i int) OperatorStep {
 func (o *Operator) Check(region *core.RegionInfo) OperatorStep {
 	for step := atomic.LoadInt32(&o.currentStep); int(step) < len(o.steps); step++ {
 		if o.steps[int(step)].IsFinish(region) {
+			operatorStepDuration.WithLabelValues(reflect.TypeOf(o.steps[int(step)]).Name()).
```
Comment (siddontang) [inline server/schedule/operator.go]: can we avoid reflect here? seem strange here.

**Pull request merged**